rpc/src/v1/helpers/dispatch.rs

```diff
@@ -523,32 +586,62 @@ impl Future for ProspectiveSigner {
 					match self.poll_reserved()? {
 						Async::NotReady => {
 							self.state = WaitForNonce;
-							self.prospective = Some(self.sign(self.reserved.prospective_value()));
+							self.prospective = Some(self.sign(self.reserved.prospective_value())?);
 						},
 						Async::Ready(nonce) => {
-							self.state = Finish;
-							self.prospective = Some(self.sign(nonce.value()));
+							self.state = PostSign;
+							self.prospective = Some(self.sign(nonce.value())?);
 							self.ready = Some(nonce);
 						},
 					}
 				},
 				WaitForNonce => {
 					let nonce = try_ready!(self.poll_reserved());
-					let result = match (self.prospective.take(), nonce.matches_prospective()) {
+					let prospective = match (self.prospective.take(), nonce.matches_prospective()) {
 						(Some(prospective), true) => prospective,
-						_ => self.sign(nonce.value()),
+						_ => self.sign(nonce.value())?,
 					};
-					self.state = Finish;
-					self.prospective = Some(result);
+					self.state = PostSign;
+					self.prospective = Some(prospective);
 					self.ready = Some(nonce);
 				},
-				Finish => {
+				PostSign => {
 					if let (Some(result), Some(nonce)) = (self.prospective.take(), self.ready.take()) {
-						// Mark nonce as used on successful signing
-						return result.map(move |tx| {
-							nonce.mark_used();
-							Async::Ready(tx)
-						})
+						let mut post_sign_future = self.post_sign.take()
+							.expect("post_sign is set on creation; qed")
+							.execute(result)
+							.into_future();
+
+						match post_sign_future.poll()? {
```

Comment:
Can we just populate `self.post_sign_future` and then `loop` again so that the logic doesn't need to be duplicated?